06_gpu_and_ml/stable_lm/main.py

```diff
@@ -0,0 +1,146 @@
+        gpu="A10G"
+    )
+    .run_function(
+        fetch_models,
+        gpu=None,
+        timeout=3600,
+    )
+    .run_function(
+        compile_models,
+        gpu="A10G",
+        timeout=3600,
+    )
+)
+
+stub = modal.Stub(name="example-stability-lm", image=image)
+
+
+class StopOnTokens(StoppingCriteria):
+    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor, **kwargs) -> bool:
+        stop_ids = [50278, 50279, 50277, 1, 0]
+        for stop_id in stop_ids:
+            if input_ids[0][-1] == stop_id:
+                return True
+        return False
+
+
+@stub.cls(gpu="A10G")
+class StabilityLM:
+    def __init__(self, model_url: str = "stabilityai/stablelm-tuned-alpha-7b"):
+        self.model_url = model_url
+
+    def __enter__(self):
+        """
+        Container-lifeycle method for model setup.
+        """
+        import accelerate
+        from transformers import AutoModelForCausalLM, AutoTokenizer
+
+        start = time.time()
+        tokenizer = AutoTokenizer.from_pretrained(self.model_url)
+        model = AutoModelForCausalLM.from_pretrained(
+            self.model_url,
+            torch_dtype=torch.float16,
+            device_map="auto",
+        )
+        print(f"Loaded model in {time.time() - start:.2f} seconds.")
+
+        if hasattr(model.config, "max_sequence_length"):
+            context_len = model.config.max_sequence_length
+        else:
+            context_len = 4096
+
+        self.context_len = context_len
+        self.generator = pipeline(
+            "text-generation",
+            model=model,
+            tokenizer=tokenizer,
+            torch_dtype=torch.float16,
+            device_map="auto",
+        )
+
+    @modal.method()
+    def generate(
+        self, text: str, temperature: float = 1.0, max_new_tokens: int = 1024
+    ):
+        text = format_prompt(text)
+        result = self.generator(
+            text,
+            temperature=temperature,
+            max_new_tokens=max_new_tokens,
+            num_return_sequences=1,
+            num_beams=1,
+            do_sample=True,
+            top_p=0.95,
+            top_k=1000,
+            stopping_criteria=StoppingCriteriaList([StopOnTokens()]),
+        )
+        return {"response": result[0]["generated_text"].replace(text, "")}
+
+
+def format_prompt(instruction):
+    return f"<|USER|>{instruction}<|ASSISTANT|>"
+
+
+@stub.function()
+@modal.web_endpoint(method="GET")
+def ask(prompt: str, temperature: float = 1.0, max_new_tokens: int = 512):
+    return StabilityLM().generate.call(
+        prompt, temperature=temperature, max_new_tokens=max_new_tokens
+    )
+
+
+@stub.local_entrypoint()
+def main():
+    q_style, q_end = "[bold bright_magenta]", "[/bold bright_magenta]"
+    instructions = [
+        "Generate a list of 20 great names for sentient cheesecakes that teach SQL.",
```

Comment:
```suggestion
        "Generate a list of the 10 most beautiful cities in the world..",
```

StableLM does very poorly on this cheesecake prompt compared to GPT-4 😄. 

```
1. Liam
2. Mila
3. Jadine
4. Pike
5. Olive
...
```

Made another suggestion, which is easier and where the answer seems more reasonable, though I'd hardly count NYC as the 3rd most _beautiful_ city in the world.